pkg/controllers/clusterresourceplacement/clusterresourceplacement_controller.go

```diff
@@ -198,15 +308,79 @@ func (r *Reconciler) lookupLatestClusterPolicySnapshot(ctx context.Context, crp
 	return &snapshotList.Items[index], lastPolicyIndex, nil
 }
 
+// lookupLatestResourceSnapshot finds the latest snapshots and its resource index.
+// There will be only one active resource snapshot if exists.
+// It first checks whether there is an active resource snapshot.
+// If not, it finds the one whose resourceIndex label is the largest.
+// The resource index will always start from 0.
+// Return error when 1) cannot list the snapshots 2) there are more than one active resource snapshots 3) snapshot has the
+// invalid label value.
+// 2 & 3 should never happen.
+func (r *Reconciler) lookupLatestResourceSnapshot(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) (*fleetv1beta1.ClusterResourceSnapshot, int, error) {
+	snapshotList := &fleetv1beta1.ClusterResourceSnapshotList{}
+	latestSnapshotLabelMatcher := client.MatchingLabels{
+		fleetv1beta1.CRPTrackingLabel:      crp.Name,
+		fleetv1beta1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+	}
+	crpKObj := klog.KObj(crp)
+	if err := r.Client.List(ctx, snapshotList, latestSnapshotLabelMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list active clusterResourceSnapshots", "clusterResourcePlacement", crpKObj)
+		return nil, -1, controller.NewAPIServerError(false, err)
+	}
+	if len(snapshotList.Items) == 1 {
```

Comment:
do we handle the multiple resourceSnapshots objects per index?

Reply:
not now, will handle it in the second phase.